include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp

```diff
@@ -356,29 +362,46 @@ class stream_ordered_memory_resource : public crtp<PoolResource>, public device_
    * @return A block with non-null pointer and size >= `size`, or a nullptr block if none is
    *         available in `blocks`.
    */
-  block_type get_block_from_other_stream(size_t size, stream_event_pair stream_event)
+  block_type get_block_from_other_stream(size_t size,
+                                         stream_event_pair stream_event,
+                                         free_list& blocks,
+                                         bool merge_first)
   {
-    // nothing in this stream's free list, look for one on another stream
-    for (auto s = stream_free_blocks_.begin(); s != stream_free_blocks_.end(); ++s) {
-      auto blocks_event = s->first;
-      if (blocks_event.event != stream_event.event) {
-        auto blocks = s->second;
-
-        block_type const b = blocks.get_block(size);  // get the best fit block
+    for (auto it = stream_free_blocks_.begin(), next_it = it; it != stream_free_blocks_.end();
+         it = next_it) {
+      ++next_it;
```

Comment:
You can use a post increment in the `for` loop here:
```suggestion
         it = next_it++) {
```

Reply:
Making this change causes segfaults in some cases. It took me a while, but now I understand why the change you suggest is not semantically the same as the current loop.

With the current loop, as soon as you enter the loop, `next_it` points one element past `it`. This is important so that we can delete `it` and still proceed with the loop. With the change you suggest, `it` and `next_it` refer to the same element of the list until after the loop body is executed (after `it` may have been deleted). We can't fix this by initializing `next_it` to the element after `begin()` because we first need to check if `begin() == end()`.

I just now looked at using an alternative structure (while loop) and taking advantage of the return value of `map::erase()`, but that resulted in a more complex structure where the iterator increment needed to be done in multiple places. The current structure allows the `next_it` logic to all be at the top of the loop.

Reply:
That said, your comment did help me find a minor bug (perhaps not a functionality bug, maybe a performance issue) in the function that calls this one.

Reply:
Might be helpful to add a comment reflecting that info
>next_it points one element past it. This is important so that we can delete it and still proceed with the loop

Reply:
Good idea, added.